README.md
# benchmark-primitive

A quick benchmark of two types in Haskell's `primitve` library.

Have two implemenations of a WordArray. Once builds on top of `ByteArray` and
the other uses `Array Word`. I expected the performance of these two
implementations to be identical, but for some reason the `Array Word` one
seems to be roughly 4 times slower than the `ByteArray` version.

I am open to the possibilty of this being a bug in my code. If its not, then
`Data.Primitve` needs fixing.

To build and run the benchmark:
```
cabal sandbox init
cabal install --dependencies-only
cabal build
dist/build/benchmark-primitive/benchmark-primitive
```

Running this test with ghc-8.0.1 on a recent Macbook Pro running Debian
Linux, I'm getting results like this:
```
benchmarking copyWordArray 1000/Data.Primitive,Array Word
time                 3.438 μs   (3.432 μs .. 3.444 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 3.437 μs   (3.434 μs .. 3.448 μs)
std dev              20.61 ns   (10.54 ns .. 40.85 ns)

benchmarking copyWordArray 1000/Data.Primitive.ByteArray
time                 612.2 ns   (611.3 ns .. 613.2 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 611.9 ns   (611.2 ns .. 613.1 ns)
std dev              2.910 ns   (1.635 ns .. 4.961 ns)

benchmarking copyWordArray 10000/Data.Primitive,Array Word
time                 28.71 μs   (28.63 μs .. 28.79 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 28.67 μs   (28.63 μs .. 28.75 μs)
std dev              193.3 ns   (117.3 ns .. 346.1 ns)

benchmarking copyWordArray 10000/Data.Primitive.ByteArray
time                 6.001 μs   (5.992 μs .. 6.012 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 6.000 μs   (5.994 μs .. 6.014 μs)
std dev              30.21 ns   (18.17 ns .. 54.13 ns)

benchmarking copyWordArray 100000/Data.Primitive,Array Word
time                 291.5 μs   (291.0 μs .. 292.2 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 291.8 μs   (291.3 μs .. 292.8 μs)
std dev              2.070 μs   (1.146 μs .. 3.810 μs)

benchmarking copyWordArray 100000/Data.Primitive.ByteArray
time                 60.88 μs   (60.86 μs .. 60.90 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 60.93 μs   (60.91 μs .. 61.00 μs)
std dev              110.9 ns   (42.01 ns .. 212.5 ns)

```






ChangeLog.md
# Revision history for benchmark-primitive

## 0.1.0.0  -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


Bench/WordArray.hs
{-# LANGUAGE BangPatterns #-}
module Bench.WordArray where

import Control.Monad.Primitive
import Data.Primitive.Array


type WordArray = Array Word

type MutableWordArray m = MutableArray (PrimState m) Word

{-# INLINE newWordArray #-}
newWordArray :: PrimMonad m => Int -> m (MutableWordArray m)
newWordArray !len = newArray len 0

{-# INLINE readWordArray #-}
readWordArray :: PrimMonad m => MutableWordArray m -> Int -> m Word
readWordArray !marr !i = readArray marr i

{-# INLINE unsafeFreezeWordArray #-}
unsafeFreezeWordArray :: PrimMonad m => MutableWordArray m -> m WordArray
unsafeFreezeWordArray !marr = unsafeFreezeArray marr

{-# INLINE indexWordArray #-}
indexWordArray :: WordArray -> Int -> Word
indexWordArray !arr !i = indexArray arr i

{-# INLINE indexWordArrayM #-}
indexWordArrayM :: Monad m => WordArray -> Int -> m Word
indexWordArrayM !arr !i = case indexArray arr i of x -> return x

{-# INLINE writeWordArray #-}
writeWordArray :: PrimMonad m => MutableWordArray m -> Int -> Word -> m ()
writeWordArray !marr !i !w = writeArray marr i w

{-# INLINE setWordArray #-}
setWordArray :: PrimMonad m => MutableWordArray m -> Int -> Int -> Word -> m ()
setWordArray !marr !offset !count !word =
    loop offset (offset + count)
  where
    loop !off !end
        | off < end = do
            writeWordArray marr off word
            loop (off + 1) end
        | otherwise = return ()

{-# INLINE copyWordArray #-}
copyWordArray :: PrimMonad m => MutableWordArray m -> Int -> WordArray -> Int -> Int -> m ()
copyWordArray !marr !doff !arr !soff !wrds =
    let loop !i
            | i < wrds =  do
                !x <- indexWordArrayM arr (soff + i)
                writeWordArray marr (doff + i) x
                loop (i + 1)
            | otherwise = return ()
    in loop 0


Bench/WordArrayB.hs
{-# LANGUAGE BangPatterns #-}
module Bench.WordArrayB where

import Control.Monad.Primitive
import Data.Primitive


newtype WordArray = WA ByteArray

newtype MutableWordArray m = MWA (MutableByteArray (PrimState m))

{-# INLINE newWordArray #-}
newWordArray :: PrimMonad m => Int -> m (MutableWordArray m)
newWordArray !len = do
    !marr <- newByteArray (len * sizeOf (0 :: Word))
    return $ MWA marr

{-# INLINE readWordArray #-}
readWordArray :: PrimMonad m => MutableWordArray m -> Int -> m Word
readWordArray (MWA !marr) = readByteArray marr

{-# INLINE unsafeFreezeWordArray #-}
unsafeFreezeWordArray :: PrimMonad m => MutableWordArray m -> m WordArray
unsafeFreezeWordArray (MWA !marr) = do
    !arr <- unsafeFreezeByteArray marr
    return (WA arr)

{-# INLINE indexWordArray #-}
indexWordArray :: WordArray -> Int -> Word
indexWordArray (WA !arr) = indexByteArray arr

{-# INLINE indexWordArrayM #-}
indexWordArrayM :: Monad m => WordArray -> Int -> m Word
indexWordArrayM (WA !arr) !i = case indexByteArray arr i of x -> return x

{-# INLINE writeWordArray #-}
writeWordArray :: PrimMonad m => MutableWordArray m -> Int -> Word -> m ()
writeWordArray (MWA !marr) = writeByteArray marr

{-# INLINE setWordArray #-}
setWordArray :: PrimMonad m => MutableWordArray m -> Int -> Int -> Word -> m ()
setWordArray !marr !offset !count !word =
    loop offset (offset + count)
  where
    loop !off !end
        | off < end = do
            writeWordArray marr off word
            loop (off + 1) end
        | otherwise = return ()

{-# INLINE copyWordArray #-}
copyWordArray :: PrimMonad m => MutableWordArray m -> Int -> WordArray -> Int -> Int -> m ()
copyWordArray !marr !doff !arr !soff !wrds =
    let loop !i
            | i < wrds =  do
                !x <- indexWordArrayM arr (soff + i)
                writeWordArray marr (doff + i) x
                loop (i + 1)
            | otherwise = return ()
    in loop 0


benchmark-primitive.hs
import qualified Bench.WordArray as A
import qualified Bench.WordArrayB as B

import qualified Criterion.Main as C

main :: IO ()
main = do
    let maxlen = 100000

    asrc <- A.unsafeFreezeWordArray =<< A.newWordArray maxlen
    adest <- A.newWordArray maxlen

    bsrc <- B.unsafeFreezeWordArray =<< B.newWordArray maxlen
    bdest <- B.newWordArray maxlen

    C.defaultMain
        [ C.bgroup "copyWordArray 1000"
            [ C.bench "Data.Primitive,Array Word"   $ C.whnfIO $ benchA 1000 asrc adest
            , C.bench "Data.Primitive.ByteArray"    $ C.whnfIO $ benchB 1000 bsrc bdest
            ]
        , C.bgroup "copyWordArray 10000"
            [ C.bench "Data.Primitive,Array Word"   $ C.whnfIO $ benchA 10000 asrc adest
            , C.bench "Data.Primitive.ByteArray"    $ C.whnfIO $ benchB 10000 bsrc bdest
            ]
        , C.bgroup "copyWordArray 100000"
            [ C.bench "Data.Primitive,Array Word"   $ C.whnfIO $ benchA 100000 asrc adest
            , C.bench "Data.Primitive.ByteArray"    $ C.whnfIO $ benchB 100000 bsrc bdest
            ]
        ]

benchA :: Int -> A.WordArray -> A.MutableWordArray IO -> IO ()
benchA len src dest = A.copyWordArray dest 0 src 0 len


benchB :: Int -> B.WordArray -> B.MutableWordArray IO -> IO ()
benchB len src dest = B.copyWordArray dest 0 src 0 len
